### PR TITLE
[AZ-638]: Implement new update strategy to devnet

### PR DESCRIPTION
**Pull request by bartoszjedrzejewski**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Infrastructure updated accordingly
- [ ] I have made corresponding changes to the documentation
- [ ] New documentation created
- [ ] Bump `spec_version` and `transaction_version` if relevant
- [ ] Bump `aleph-client` version if relevant

### CODE REVIEW

File: .github/workflows/deploy-to-devnet.yml [hunk 1]
```diff
@@ -60,21 +50,73 @@ jobs:
...
+          COMMIT_ID=${TMP_COMMIT_ID: -7}
+          echo $COMMIT_ID
+
+          if aws s3api head-object --bucket alephzero-devnet-eu-central-1-keys-bucket --key forks/chainspec-${COMMIT_ID}.json; then
+            echo "Chainspec exists"
+            IF_CHAINSPEC_EXISTS=true
+            aws s3 cp s3://alephzero-devnet-eu-central-1-keys-bucket/forks/chainspec-${COMMIT_ID}.json chainspec.skeleton.json
+          else
+            echo "Chainspec doesn't exist"
+            IF_CHAINSPEC_EXISTS=false
```
Comment (fbielejec) [inline .github/workflows/deploy-to-devnet.yml:79]: why set this if not used anywhere?

File: .github/workflows/deploy-to-devnet.yml [hunk 2]
```diff
@@ -60,21 +50,73 @@ jobs:
...
+
+            docker run -it --env RUST_BACKTRACE=1 --entrypoint "/usr/local/bin/aleph-node" public.ecr.aws/p6e8q1z1/aleph-node:${COMMIT_ID} bootstrap-chain --raw --base-path /data --chain-id a0dnet1 --account-ids 5D34dL5prEUaGNQtPPZ3yN5Y6BnkfXunKXXz6fo7ZJbLwRRH,5GBNeWRhZc2jXu7D55rBimKYDk8PGk8itRYFTPfC8RJLKG5o,5Dfis6XL8J2P6JHUnUtArnFWndn62SydeP8ee8sG2ky9nfm9,5F4H97f7nQovyrbiq4ZetaaviNwThSVcFobcA5aGab6167dK,5DiDShBWa1fQx6gLzpf3SFBhMinCoyvHM1BWjPNsmXS8hkrW,5EFb84yH9tpcFuiKUcsmdoF7xeeY3ajG1ZLQimxQoFt9HMKR,5DZLHESsfGrJ5YzT3HuRPXsSNb589xQ4Unubh1mYLodzKdVY,5GHJzqvG6tXnngCpG7B12qjUvbo5e4e9z8Xjidk3CQZHxTPZ,5CUnSsgAyLND3bxxnfNhgWXSe9Wn676JzLpGLgyJv858qhoX,5CVKn7HAZW1Ky4r7Vkgsr7VEW88C2sHgUNDiwHY9Ct2hjU8q --sudo-account-id 5F4SvwaUEQubiqkPF8YnRfcN77cLsT2DfG4vFeQmSXNjR7hD > chainspec.skeleton.json
```
Comment (fbielejec) [inline .github/workflows/deploy-to-devnet.yml:81]: maybe you could leave a comment here that these accounts are:
```
Alice Bob Charlie Dave Eve Ferdie Alice//stash Bob//stash Charlie//stash Dave//stash Eve//stash Ferdie//stash
```

File: .github/workflows/deploy-to-devnet.yml [hunk 3]
```diff
@@ -60,21 +50,73 @@ jobs:
...
 
+            aws s3 cp chainspec.skeleton.json s3://alephzero-devnet-eu-central-1-keys-bucket/forks/chainspec-${COMMIT_ID}.json
+          fi
+
+          docker run -it -v $(pwd):/app public.ecr.aws/p6e8q1z1/fork-off:latest --http-rpc-endpoint=https://rpc.test.azero.dev --initial-spec-path=chainspec.skeleton.json --combined-spec-path=chainspec.json
```
Comment (fbielejec) [inline .github/workflows/deploy-to-devnet.yml:86]: Add:
--pallets-keep-state=Aura,Aleph,Sudo,Staking,Session,Elections

These are the defaults but it never hurts to be verbose, and chances are there will be soon more pallets joining, with their own state

**Review by fbielejec — COMMENTED** [review]
Two minor comments

It is important to note, that with this setup the forked devnet will always use the same genesis state as the starting point (because it pulls it from S3). 

I'm wondering if there might be a situation when this is **not** a wanted behavior. What comes to mind is if there was a change enacted in one of the "pallets-keeps-state" - i.e. sudo account was changed, or a set of validators, or staking parameters etc. Than this change will not  get carried over to the fork.

File: .github/workflows/deploy-to-devnet.yml (hunk 3 above)
Comment (bartoszjedrzejewski) [reply .github/workflows/deploy-to-devnet.yml:86]: If there will be more pallets joining I'd rather do it by updating fork-off, which would automatically update this workflow(it uses latest image of fork-off tool). So leaving it as default is beneficial.

File: .github/workflows/deploy-to-devnet.yml (hunk 1 above)
Comment (bartoszjedrzejewski) [reply .github/workflows/deploy-to-devnet.yml:79]: leftover, thanks

**Comment by bartoszjedrzejewski** [comment]
> Two minor comments
> 
> It is important to note, that with this setup the forked devnet will always use the same genesis state as the starting point (because it pulls it from S3).
> 
> I'm wondering if there might be a situation when this is **not** a wanted behavior. What comes to mind is if there was a change enacted in one of the "pallets-keeps-state" - i.e. sudo account was changed, or a set of validators, or staking parameters etc. Than this change will not get carried over to the fork.


> Two minor comments
> 
> It is important to note, that with this setup the forked devnet will always use the same genesis state as the starting point (because it pulls it from S3).
> 
> I'm wondering if there might be a situation when this is **not** a wanted behavior. What comes to mind is if there was a change enacted in one of the "pallets-keeps-state" - i.e. sudo account was changed, or a set of validators, or staking parameters etc. Than this change will not get carried over to the fork.

You are right. But if they were to change they would have to be changed in the workflow anyway, so we can wait with this change if it will be needed(maybe not :) ).

**Review by skrolikowski-10c — APPROVED** [review]

**Pull request merged**